codec-http/src/main/java/io/netty/handler/codec/http/HttpContentDecoder.java

```diff
@@ -73,76 +72,73 @@ protected void decode(ChannelHandlerContext ctx, HttpObject msg, List<Object> ou
-                                contentLength += ((HttpContent) o).content().readableBytes();
-                            }
-                        }
-                        headers.set(
-                                HttpHeaderNames.CONTENT_LENGTH,
-                                Integer.toString(contentLength));
-                    }
-                    return;
-                }
+            // Determine the content encoding.
+            String contentEncoding = headers.getAndConvert(HttpHeaderNames.CONTENT_ENCODING);
+            if (contentEncoding != null) {
+                contentEncoding = contentEncoding.trim();
+            } else {
+                contentEncoding = IDENTITY;
+            }
+            decoder = newContentDecoder(contentEncoding);
 
-                if (c instanceof LastHttpContent) {
-                    decodeStarted = false;
+            if (decoder == null) {
+                if (message instanceof HttpContent) {
+                    ((HttpContent) message).retain();
                 }
                 out.add(message);
-                out.add(c.retain());
                 return;
             }
 
-            if (decoder != null) {
-                decodeContent(c, out);
+            // Remove content-length header:
+            // the correct value can be set only after all chunks are processed/decoded.
+            // If buffering is not an issue, add HttpObjectAggregator down the chain, it will set the header.
+            // Otherwise, rely on LastHttpContent message.
+            headers.remove(HttpHeaderNames.CONTENT_LENGTH);
+
+            // set new content encoding,
+            CharSequence targetContentEncoding = getTargetContentEncoding(contentEncoding);
+            if (HttpHeaderValues.IDENTITY.equals(targetContentEncoding)) {
+                // Do NOT set the 'Content-Encoding' header if the target encoding is 'identity'
+                // as per: http://tools.ietf.org/html/rfc2616#section-14.11
+                headers.remove(HttpHeaderNames.CONTENT_ENCODING);
             } else {
-                if (c instanceof LastHttpContent) {
-                    decodeStarted = false;
+                headers.set(HttpHeaderNames.CONTENT_ENCODING, targetContentEncoding);
+            }
+
+            if (message instanceof HttpContent) {
+                // If message is a full request or response object (headers + data), don't copy data part into out.
+                // Output headers only; data part will be decoded below.
+                // Note: "copy" object must not be an instance of LastHttpContent class,
+                // as this would (erroneously) indicate the end of the HttpMessage to other handlers.
+                HttpMessage copy;
+                if (message instanceof HttpRequest) {
+                    HttpRequest r = (HttpRequest) message; // HttpRequest or FullHttpRequest
+                    copy = new DefaultHttpRequest(r.protocolVersion(), r.method(), r.uri());
```

Comment:
hi Schott, FullHttpRequest.copy(null) cannot be used here.
it does create a Request object with headers and no data (that's good), but the returned object also implements interface LastHttpContent, which indicates the end of the request and its chunks (and this is wrong). The real end of the request is several chunks (HttpContent messages) later.